rfcs/0000_hypotheses_resampling.md

```diff
@@ -0,0 +1,75 @@
+* **Accurate resampling**:  If we sample new hypotheses close to the hypotheses with high total accumulated evidence (e.g., particle filter), we could be sampling from the incorrect hypotheses (if we had just switched objects). If we sample close to the hypotheses with high evidence slope, we may converge faster.
+* **Practical**: The total evidence can still be unbounded, it doesn't matter because we only consider the slope. This metric does not care about how much evidence we've accumulated already. It other words, a hypothesis with a high evidence, can be removed if it hasn't accumulated evidence in a while, while a consistently growing hypothesis is less likely to be removed even if it was just added.
+
+## Assumptions and constraints:
+* Number of hypotheses should not scale up with steps, if anything they should decrease. For now, any sampled hypothesis must replace an old "unlikely" hypothesis.
+* Sampling of likely and unlikely hypotheses is based on evidence change instead of absolute evidence.
+* Terminal state calculation is still based on accumulated evidence and `x_percent_threshold`.
+* The resampling procedure only occurs if principal curvature is defined.
+
+## The Resampling Procedure
+
+1) For every object, at every new step, get initial hypotheses based on the observed pose.
+2) If principal curvature is not defined, skip
+3) Uniformly sample M percent from these new hypotheses (e.g., M = 20% = 1000 hyp)
+4) Sample N percent from the existing hypotheses distribution of highest evidence slope (e.g., N = 10% = 500 hyp). These sampled hypotheses should be "close" to the most likely hypotheses.
+5) Replace unlikely M+N percent of existing hypotheses (e.g., 30% = 1500 hyp) with these newly sampled hypotheses. The unlikely hypotheses are chosen based on evidence change.
+
+# Alternatives Considered:
+* Detect change with evidence decay and hard reset?
+	* Here we would focus on detecting the change that triggers full reset of Monty states. 
+	* Detecting the object change will likely be based on evidence decay and can be problematic. If an object was added and quickly changed before accumulating enough evidence, it will be difficult to detect change based on evidence decay. This will still result in incorrect initial hypotheses. Resampling without a hard reset is more general.
+
+# Future possibilities
+* This only works in inference assuming that we have good object models learned in isolation. It likely breaks learning from scratch experiments.
+* I manually force Monty back into the matching phase and reset it's counters between episodes. 
+	* Maybe we should brainstorm about what a terminal state should be in a multi-object environment.
+	* Maybe also brainstorm about ways to more seamlessly switch between matching and exploration? I don't like forcing Monty to always be in matching mode, by doing this it feels that we are making the gap between learning and inference wider.
```

Comment:
Definitely two good topics to talk about more. I wrote my thoughts on both in earlier comments. lmk what you think.